src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java

```diff
@@ -0,0 +1,393 @@
+     * @param metisCrudAction the action that was performed
+     * @param conversation    the conversation that was affected
+     * @param usersToMessage  the users to be messaged
+     */
+    public void broadcastOnConversationMembershipChannel(Course course, MetisCrudAction metisCrudAction, Conversation conversation, Set<User> usersToMessage) {
+        String conversationParticipantTopicName = getConversationParticipantTopicName(course.getId());
+        usersToMessage.forEach(user -> sendToConversationMembershipChannel(metisCrudAction, conversation, user, conversationParticipantTopicName));
+    }
+
+    @NotNull
+    public static String getConversationParticipantTopicName(Long courseId) {
+        return METIS_WEBSOCKET_CHANNEL_PREFIX + "courses/" + courseId + "/conversations/user/";
+    }
+
+    private void sendToConversationMembershipChannel(MetisCrudAction metisCrudAction, Conversation conversation, User user, String conversationParticipantTopicName) {
+        ConversationDTO dto;
+        if (metisCrudAction.equals(MetisCrudAction.NEW_MESSAGE)) {
+            // we do not want to recalculate the whole dto for a new message, just the information needed for updating the unread messages
+            dto = conversationDTOService.convertToDTOWithoutExtraDBCalls(conversation);
+        }
+        else {
+            dto = conversationDTOService.convertToDTO(conversation, user);
+        }
+
+        var websocketDTO = new ConversationWebsocketDTO(dto, metisCrudAction);
+        messagingTemplate.convertAndSendToUser(user.getLogin(), conversationParticipantTopicName + user.getId(), websocketDTO);
+    }
+
+    /**
+     * Checks if a user is a member of a conversation and therefore can access it else throws an exception
+     *
+     * @param conversationId the id of the conversation
+     * @param user           the user to check
+     * @return conversation if the user is a member
+     */
+    public Conversation mayInteractWithConversationElseThrow(Long conversationId, User user) {
+        Optional<Conversation> conversation = conversationRepository.findById(conversationId);
+        if (conversation.isEmpty() || !isMember(conversationId, user.getId())) {
+            throw new AccessForbiddenException("User not allowed to access this conversation!");
+        }
+        return conversation.get();
+    }
+
+    /**
+     * Updates the last read date time of a user for a conversation
+     *
+     * @param conversation the conversation
+     * @param user         the user
+     * @return the new last read date time
+     */
+    public ZonedDateTime auditConversationReadTimeOfUser(Conversation conversation, User user) {
+        // update the last time user has read the conversation
+        ConversationParticipant readingParticipant = conversationParticipantRepository.findConversationParticipantByConversationIdAndUserId(conversation.getId(), user.getId())
```

Comment:
The method for `orElseThrow` should be in the Repository.

Reply:
Done in 1db0a029fc229bb2463a3fae195117c3a1f3aa81